src/System.Private.ServiceModel/src/System/ServiceModel/Channels/HttpResponseMessageHelper.cs

```diff
@@ -0,0 +1,367 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Diagnostics.Contracts;
+using System.IO;
+using System.Net;
+using System.Net.Http;
+using System.ServiceModel.Diagnostics.Application;
+using System.ServiceModel.Security;
+using System.Threading.Tasks;
+using System.Xml;
+
+namespace System.ServiceModel.Channels
+{
+    internal class HttpResponseMessageHelper
+    {
+        private readonly HttpChannelFactory<IRequestChannel> _factory;
+        private readonly MessageEncoder _encoder;
+        private readonly HttpRequestMessage _httpRequestMessage;
+        private readonly HttpResponseMessage _httpResponseMessage;
+        private string _contentType;
+        private long _contentLength;
+
+        public HttpResponseMessageHelper(HttpResponseMessage httpResponseMessage, HttpChannelFactory<IRequestChannel> factory)
+        {
+            Contract.Assert(httpResponseMessage != null);
+            Contract.Assert(httpResponseMessage.RequestMessage != null);
+            _httpResponseMessage = httpResponseMessage;
+            _httpRequestMessage = httpResponseMessage.RequestMessage;
+            _factory = factory;
+            _encoder = factory.MessageEncoderFactory.Encoder;
+        }
+
+        internal async Task<Message> ParseIncomingResponse()
+        {
+            ValidateAuthentication();
+            ValidateResponseStatusCode();
+            bool hasContent = await ValidateContentTypeAsync();
+            Message message = null;
+
+            if (!hasContent)
+            {
+                if (_encoder.MessageVersion == MessageVersion.None)
+                {
+                    message = new NullMessage();
+                }
+                else
+                {
+                    return null;
+                }
+            }
+            else
+            {
+                message = await ReadStreamAsMessageAsync();
+            }
+
+            var exception = ProcessHttpAddressing(message);
+            Contract.Assert(exception == null, "ProcessHttpAddressing should not set an exception after parsing a response message.");
+
+            return message;
+        }
+
+        private Exception ProcessHttpAddressing(Message message)
+        {
+            Exception result = null;
+            AddProperties(message);
+
+            // check if user is receiving WS-1 messages
+            if (message.Version.Addressing == AddressingVersion.None)
+            {
+                bool actionAbsent = false;
+                try
+                {
+                    actionAbsent = (message.Headers.Action == null);
```

Comment:
I don't know that `Message.Headers` guaranteed to be non-null (it's abstract, so it really depends on the implementer). 
This would be a really weird place to null ref if it happened.